Remove json data attribute

leungkinghin opened the pull request:
- Remove data attribute inside the JSON object returning from payment handler https://github.com/commercetools/commercetools-adyen-integration/pull/630#discussion_r600740392

Original
```
{
    success : true,
    data : {
        actions : [ { ... } ]   
    }
}
```
```
{
    success : false,
    data : {
        errors : [ { ... } ]   
    }
}
```
Now
```
{
    success : true,
    actions : [ { ... } ]
}
```
```
{
    success : false,
    errors : [ { ... } ]
}
```
- Modify payment controller
- Update corresponding integration test / unit test

lojzatran commented on extension/index.googleFunction.js:
```diff
@@ -22,11 +22,11 @@ exports.extensionTrigger = async (request, response) => {
     if (paymentResult.success) {
       return response.status(200).send({
-        actions: paymentResult.data ? paymentResult.data.actions : [],
+        actions: paymentResult.actions ? paymentResult.actions : [],
```
```suggestion
        actions: paymentResult.actions || [],
```

leungkinghin replied:
https://github.com/commercetools/commercetools-adyen-integration/pull/656/commits/697e5ef85b3a601780e043a73f6bd34281ae22cb

leungkinghin replied:
https://github.com/commercetools/commercetools-adyen-integration/pull/656/commits/2415e2a967e45e1488c4800a3354ae836ff9194e

lojzatran commented on extension/src/paymentHandler/payment-handler.js:
```diff
@@ -122,21 +119,21 @@ function _validatePaymentRequest(paymentObject, authToken) {
       return {
         success: false,
-        data: paymentValidator.buildCtpErrorResponse(),
+        errors: paymentValidator.buildCtpErrorResponse()?.errors,
```
Do you need the null check here again? Seems like you check already `paymentValidator.hasErrors()`?

leungkinghin replied:
you're right. `errors` inside `paymentValidator` is always non-null
https://github.com/commercetools/commercetools-adyen-integration/pull/656/commits/64058c6cac27e381313dc82dee9149fd8f7a3d80

lojzatran replied:
You have this check in multiple places.

leungkinghin replied:
https://github.com/commercetools/commercetools-adyen-integration/pull/656/commits/b34bccc36c56fe0f323e5bd457a8ae5cc3c8aaef

lojzatran commented on extension/src/api/payment/payment.controller.js:
```diff
@@ -30,12 +30,15 @@ async function processRequest(request, response) {
-      data: paymentResult.data,
-    })
+      data: paymentResult.success
+        ? { actions: paymentResult?.actions }
```
I guess you don't need null-check here, otherwise it would fail already when you call `paymentResult.success`.

leungkinghin replied:
Just checked. actions attribute depends on what handlers return, which never contain null actions.
https://github.com/commercetools/commercetools-adyen-integration/pull/656/commits/869fa2d1bae80a77e668eb49419f72c2a50c657e

lojzatran commented:
What was the timeout issue in e2e test and how did you resolve it?

leungkinghin commented:
> What was the timeout issue in e2e test and how did you resolve it?

It is not about timeout. Changed json object and the incorrect structure causes assertion fail in e2e after make payment. I just fixed this issue.

butenkor commented on extension/src/paymentHandler/payment-handler.js:
```diff
@@ -122,21 +119,21 @@ function _validatePaymentRequest(paymentObject, authToken) {
     paymentValidator.validateMetadataFields()
     if (paymentValidator.hasErrors())
       return {
         success: false,
```
I wonder if this flag is required at all? I guess if `errors` is not empty then it has the same meaning as above, right?

leungkinghin replied:
https://github.com/commercetools/commercetools-adyen-integration/pull/656/commits/04d7a6bdceb15f77466d9008ff7b8dcfba7f2da1

lojzatran commented on extension/index.googleFunction.js:
```diff
@@ -22,11 +22,11 @@ exports.extensionTrigger = async (request, response) => {
     }
     return response.status(400).send({
-      errors: paymentResult.data ? paymentResult.data.errors : undefined,
+      errors: paymentResult?.errors,
```
```suggestion
      errors: paymentResult.errors,
```

leungkinghin replied:
https://github.com/commercetools/commercetools-adyen-integration/pull/656/commits/18eafc8eff9510648326a1c00caafa81780df517

lojzatran commented on extension/index.lambda.js:
```diff
@@ -29,9 +27,8 @@ exports.handler = async (event) => {
-      // paymentResult.data can be null when paymentHandler short-circuits on non-Adyen payment
-      errors: paymentResult.data ? paymentResult.data.errors : undefined,
-      actions: paymentResult.data ? paymentResult.data.actions : [],
+      errors: paymentResult?.errors,
```
```suggestion
      errors: paymentResult.errors,
```

leungkinghin replied:
https://github.com/commercetools/commercetools-adyen-integration/pull/656/commits/18eafc8eff9510648326a1c00caafa81780df517

lojzatran commented on extension/src/api/payment/payment.controller.js:
```diff
@@ -30,12 +30,15 @@ async function processRequest(request, response) {
-    })
+      data: paymentResult.success
+        ? { actions: paymentResult.actions }
+        : { errors: paymentResult?.errors },
```
```suggestion
        : { errors: paymentResult.errors },
```

leungkinghin replied:
https://github.com/commercetools/commercetools-adyen-integration/pull/656/commits/18eafc8eff9510648326a1c00caafa81780df517

butenkor commented on extension/src/api/payment/payment.controller.js:
```diff
@@ -30,12 +30,15 @@ async function processRequest(request, response) {
       statusCode: paymentResult.success ? 200 : 400,
-      data: paymentResult.data,
-    })
+      data: paymentResult.success
```
Do we need this flag if we can check for existence of errors instead?

leungkinghin replied:
https://github.com/commercetools/commercetools-adyen-integration/pull/656/commits/3706831713c202a1f0bc0671a40ecc4ffe52a098

butenkor commented on extension/src/paymentHandler/payment-handler.js:
```diff
@@ -26,27 +26,28 @@ const PAYMENT_METHOD_TYPE_KLARNA_METHODS = [
     // if it's not adyen payment, ignore the payment
-    return { success: true, data: null }
-
+    return { success: true, actions: [] }
```
Do we need success flag?

leungkinghin replied:
https://github.com/commercetools/commercetools-adyen-integration/pull/656/commits/3706831713c202a1f0bc0671a40ecc4ffe52a098

butenkor commented on extension/src/paymentHandler/payment-handler.js:
```diff
@@ -26,27 +26,28 @@ const PAYMENT_METHOD_TYPE_KLARNA_METHODS = [
-  if (validatePaymentErrors) return validatePaymentErrors
+  if (validatePaymentErrors)
+    return {
+      success: false,
```
As above as also below :)

leungkinghin replied:
https://github.com/commercetools/commercetools-adyen-integration/pull/656/commits/3706831713c202a1f0bc0671a40ecc4ffe52a098

lojzatran commented on extension/index.googleFunction.js:
```diff
@@ -22,11 +22,11 @@ exports.extensionTrigger = async (request, response) => {
     )
     if (paymentResult.success) {
```
Didn't you remove this `success` flag?

leungkinghin replied:
https://github.com/commercetools/commercetools-adyen-integration/pull/656/commits/c8c8588ec91d3750e5a68597667c8777624eee87

lojzatran replied:
I still see you have `success` attr in the code. Check for all occurrences of it.

leungkinghin replied:
sorry for missing
https://github.com/commercetools/commercetools-adyen-integration/pull/656/commits/23fef0bc40d6506b90fa7379dbd06236fc0d78ee

remove all success attribute in extension.

butenkor commented on extension/src/paymentHandler/payment-handler.js:
```diff
@@ -120,35 +119,23 @@ function _validatePaymentRequest(paymentObject, authToken) {
-        success: false,
-        data: paymentValidator.buildCtpErrorResponse(),
-      }
+      return paymentValidator.buildCtpErrorResponse().errors
```
Maybe instead of asking always for `.errors` the `buildCtpErrorResponse()` could be renamed to `getErrors()` which would return an error array right away?

leungkinghin replied:
https://github.com/commercetools/commercetools-adyen-integration/pull/656/commits/9bf490870b4ef7dd0d21f0f216d10fbb1bfbc529
